### PR TITLE
feat(errors): add new "errors" package

### PR DESCRIPTION
**Pull request by cjthompson**
# Description
This PR creates a new package called `errors`.

`errors` is designed to provide common named errors to make error handling more consistent. For example, in a plugin that talks to a backend API we shouldn't pass through the backend error directly. We should provide a named error and embed the backend error in the `cause` property.

For example:

```ts
try {
  await axios.get('...')
} catch (e) {
  if (e.statusCode === 429) {
    throw new new RateLimitError('Rate limited', { cause: e })
}
```

# Discussion
We could change the errors to not accept a `message` but to create a default message. For example, `ValidationError` could take `{ details: { name: 'property', expected: 'a string', actual: 'null' } }` and generate a `message` of `Property ${name} is expected to be ${expected} but received ${actual}`.

### CODE REVIEW

File: packages/errors/src/ErrorWithCause.test.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+const foo3 = new ErrorWithCause<string>('bar', {})
+const foo4 = new ErrorWithCause<string>('bar', { cause: undefined })
+const foo5 = new ErrorWithCause<string>('bar', { cause: 'bar' })
+
+const foo6 = new ErrorWithCause('baz')
+const foo7 = new ErrorWithCause('baz', {})
+const foo8 = new ErrorWithCause('baz', { cause: undefined })
+const foo9 = new ErrorWithCause('baz', { cause: 'bar' })
+
+ */
```
Comment (cjthompson) [inline packages/errors/src/ErrorWithCause.test.ts:77]:
```suggestion
```

File: packages/errors/src/ErrorWithDetails.test.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+describe('ErrorWithDetails', () => {
+  it('should create an error with no mesasge', () => {
+    const e = new ErrorWithDetails()
+    expect(e).toBeInstanceOf(ErrorWithDetails)
+    expect(e.message).toBe('')
+    expect(e.cause).toBeUndefined()
+  })
+
+  it('should create an error with a mesasge', () => {
```
Comment (cjthompson) [inline packages/errors/src/ErrorWithDetails.test.ts:11]:
```suggestion
  it('should create an error with a message', () => {
```

File: packages/errors/src/ErrorWithDetails.test.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    const e = new ErrorWithDetails('test message')
+    expect(e.message).toBe('test message')
+    expect(e.cause).toBeUndefined()
+  })
+
+  it('should create an error with a mesasge and a cause', () => {
```
Comment (cjthompson) [inline packages/errors/src/ErrorWithDetails.test.ts:17]:
```suggestion
  it('should create an error with a message and a cause', () => {
```

File: packages/errors/src/ErrorWithDetails.test.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    const cause = new TypeError('cause')
+    const e = new ErrorWithDetails('test message', { cause })
+    expect(e.message).toBe('test message')
+    expect(e.cause).toBe(cause)
+  })
+
+  it('should create an error with a mesasge and details', () => {
```
Comment (cjthompson) [inline packages/errors/src/ErrorWithDetails.test.ts:24]:
```suggestion
  it('should create an error with a message and details', () => {
```

File: packages/errors/src/errors.test.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+describe.each`
+  TestError
+  ${ErrorWithDetails}
+  ${ForbiddenError}
+  ${NotFoundError}
+  ${RateLimitError}
+  ${UnauthorizedError}
+  ${ValidationError}
+`('$TestError.name', ({ TestError }) => {
+  it('should create an error with no mesasge', () => {
```
Comment (cjthompson) [inline packages/errors/src/errors.test.ts:17]:
```suggestion
  it('should create an error with no message', () => {
```

File: packages/errors/src/errors.test.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    const e = new TestError()
+    expect(e).toBeInstanceOf(TestError)
+    expect(e.message).toBe('')
+    expect(e.cause).toBeUndefined()
+  })
+
+  it('should create an error with a mesasge', () => {
```
Comment (cjthompson) [inline packages/errors/src/errors.test.ts:24]:
```suggestion
  it('should create an error with a message', () => {
```

File: packages/errors/src/errors.test.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    const e = new TestError('test message')
+    expect(e.message).toBe('test message')
+    expect(e.cause).toBeUndefined()
+  })
+
+  it('should create an error with a mesasge and a cause', () => {
```
Comment (cjthompson) [inline packages/errors/src/errors.test.ts:30]:
```suggestion
  it('should create an error with a message and a cause', () => {
```

File: packages/errors/src/errors.test.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    const cause = new TypeError('cause')
+    const e = new TestError('test message', { cause })
+    expect(e.message).toBe('test message')
+    expect(e.cause).toBe(cause)
+  })
+
+  it('should create an error with a mesasge and details', () => {
```
Comment (cjthompson) [inline packages/errors/src/errors.test.ts:37]:
```suggestion
  it('should create an error with a message and details', () => {
```

File: packages/errors/src/ErrorWithDetails.test.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
+import ErrorWithDetails from './ErrorWithDetails'
+
+describe('ErrorWithDetails', () => {
+  it('should create an error with no mesasge', () => {
```
Comment (cjthompson) [inline packages/errors/src/ErrorWithDetails.test.ts:4]:
```suggestion
  it('should create an error with no message', () => {
```

**Review by 0xdef1cafe — APPROVED** [review]

**Review by theoboldfrazier — APPROVED** [review]

File: packages/errors/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,8 @@
+export { createErrorClass } from './createErrorClass'
+export { default as ErrorWithCause } from './ErrorWithCause'
```
Comment (0xApotheosis) [inline packages/errors/src/index.ts:2]: Do we need to have `ErrorWithCause` if we already have `ErrorWithDetails`, which can have both `cause` and `details`?

I'm not clear on the advantage of a cause-only error.

**Pull request merged**